Add TranspileLayout to the C API (#14780)
Qiskit/qiskit: docs/cdoc/qk-transpiler-layout.rst

=================
QkTranspileLayout
=================

.. code-block:: c

   typedef struct QkTranspileLayout QkTranspileLayout

> The ``QkTranspileLayout`` type is used to module the permutations introduced by

Reviewer: ```suggestion
The ``QkTranspileLayout`` type is used to model the permutations introduced by
```

the transpiler. In general Qiskit's transpiler is unitary-preserving up to the
initial layout and routing permutations. The initial layout permutation is
caused by setting and applying the initial layout (the mapping from virtual
circuit qubits to physical qubits on the target) and the routing permtuations
are caused by swap gate insertion or permutation ellision prior to the initial
layout being set in the transpiler pipeline. This type tracks these details and

> provide an interface to reason about these permutations.

Reviewer: It'd probably be good to either centralise this non-programming high-level discussion, or at least link from here to the version of it in the Python-space docs (and improve it if necessary).

=================
QkTranspileLayout
=================

.. code-block:: c

   typedef struct QkTranspileLayout QkTranspileLayout

The ``QkTranspileLayout`` type is used to module the permutations introduced by
the transpiler. In general Qiskit's transpiler is unitary-preserving up to the
initial layout and output permutations. The initial layout permutation is
caused by setting and applying the initial layout (the mapping from virtual
circuit qubits to physical qubits on the target) and the output permtuations

> are caused by swap gate insertion or permutation ellision prior to the initial

Reviewer: ```suggestion
circuit qubits to physical qubits on the target) and the output permutation
is caused by swap gate insertion or permutation elision prior to the initial
```

layout being set in the transpiler pipeline. This type tracks these details and

> provide an interface to reason about these permutations.

Reviewer: It'd be good to either link to the version of this discussion in Python space, or centralise both.

Author: I'm fine doing that, which doc would you suggest I link to?

Reviewer: tbh I'm not sure we have a really good version of it in Python space yet, so let's just leave it be.  I've had Intentions™ for a few months to write them, but never actually got it done.  I'll revisit this when I do.

> provide an interface to reason about these permutations.

Reviewer: Please can we not refer to the layout as "a permutation" in new docs?  It's a relabelling - our insistence that we treat it as a permutation is what leads to the crazy "implicit ancilla" system.

Author: Sure I can change the wording here

Author: Reworded in: https://github.com/Qiskit/qiskit/pull/14780/commits/70251e491667649ef413472f968f1dbf32bedfd9

For example if you had a circuit constructed like:

.. code-block:: c

    #include <qiskit.h>

    QkCircuit *qc = qk_circuit_new(3, 0)
    uint32_t h_qargs[1] = {0};
    qk_circuit_gate(qc, QKGate_H, h_qargs, NULL);
    uint32_t cx_0_qargs[2] = {0, 1};
    qk_circuit_gate(qc, QKGate_CX, cx_0_qargs, NULL);
    uint32_t cx_1_qargs[2] = {0, 2};
    qk_circuit_gate(qc, QKGate_CX, cx_1_qargs, NULL);

and during the layout stage the transpiler maps the virtual qubits in that
circuit to the physical circuits as:

> 
> 0 -> 2, 1 -> 1, 2 -> 0
> 

Reviewer: This presumably doesn't format nicely - perhaps:
```suggestion

.. code-block:: text
    0 -> 2, 1 -> 1, 2 -> 0

```

    uint32_t swap_qargs[2] = {1, 0};
    qk_circuit_gate(qc, QKGate_Swap, swap_qargs, NULL);
    uint32_t cx_1_qargs[2] = {2, 1};
    qk_circuit_gate(qc, QKGate_CX, cx_1_qargs, NULL);

this results in the output state of qubit 0 moving to qubit 1 and qubit 1's to qubit 0.
This results in ``qk_transpile_layout_output_permutation()`` returning the array:
``[1, 0, 2]`` to indicate this is the final position of each qubit's state after the
initial layout mapping. If no swaps or permutation ellisions were made during the
transpilation this will be a trivial array of the form ``[0, 1, 2]`` as the output state
of the qubit is not moved by the transpiler.

Then combining these two is the final layout which is for tracking the final
position of a virtual qubit in the input circuit. So from the above example (with the swap),
the ``qk_transpile_layout_final_layout()`` function would return ``[2, 0, 1]`` because
following the initial layout and then any routing permutation the final position of
qubit 0 in the input circuit is now physical qubit 2, virtual qubit 1 in the input circuit is
physical qubit 0, and virtual qubit 2 in the input circuit is physical qubit 1.

> The transpiler will also allocate ancillas qubits to the circuit if the target

Reviewer: ```suggestion
The transpiler will also allocate ancilla qubits to the circuit if the target
```
